src/com/google/enterprise/adaptor/sharepoint/SharePointAdaptor.java

```diff
@@ -2618,20 +2593,13 @@ private long addMetadata(Response response, String name, String value,
           .setPermits(permits);
     }
 
-    private void addPermitUserToAcl(int userId, Acl.Builder aclToUpdate)
+    private void addPermitUserToAcl(Integer userId, Acl.Builder aclToUpdate)
         throws IOException {
       if (userId == -1) {
         return;
       }
-      Principal principal = getMemberIdMapping().getPrincipal(userId);
-      // MemberIdMapping will have information about users with explicit
-      // permissions on SharePoint or users which are direct members of
-      // SharePoint groups. MemberIdMapping might not have information
-      // about all valid SharePoint Users. To get all valid SharePoint users
-      // under SiteCollection, use SiteUserMapping.
-      if (principal == null) {
-        principal = getSiteUserMapping().getPrincipal(userId);
-      }
+      IdMappings idMapping = new IdMappings();
+      Principal principal = idMapping.resolvePrincipal(userId);
```

Comment:
The problem with doing this here is that we reset the siteUserMappingRefreshed flag and may refresh the site user mapping for this single user (the author) right after potentially refreshing it to process the bulk of the ACL. We could ignore that, on the theory that if the site user mapping was recently refreshed, then usually the author will be found. Or we could life the construction of the IdMappings out of generateAcl, so that the same IdMappings could be passed down to this method as was used for the call to generateAcl on this same ACL. You could even just have two generateAcl overloads, one that tasks an IdMappings and one that gets a new one and calls the first, so that this call (the one just before the call to addPermitUserToAcl) doesn't make the other four more complicated and harder to read.